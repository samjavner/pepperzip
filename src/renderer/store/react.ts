import * as React from "react";
import { Commands, Dispatch, Store, Update, UseStoreOptions } from "./";

export interface Msg<M extends object, U extends Update<M>> {
    type: keyof U;
    payload?: any;
}

export function useLocalStore(log: (msg: any) => void = () => ({})) {
    return function actual<
        M extends object,
        U extends Update<M>,
        C extends Commands = {}
    >({
        init,
        update,
        createCommands,
        memo = [],
    }: UseStoreOptions<M, U, C>): Store<M, U, C> {
        const [model, reducerDispatch] = React.useReducer(
            (state: M, action: Msg<M, U>) =>
                update[action.type](state, action.payload),
            init
        );

        const vm = React.useMemo(
            () => {
                const dispatch: Dispatch<U> = {} as any;
                Object.keys(update).forEach(type => {
                    (dispatch as any)[type] = (payload: any) => {
                        const msg =
                            update[type].length === 1
                                ? { type }
                                : { type, payload };
                        log(msg);
                        reducerDispatch(msg);
                    };
                });
                const commands: C = createCommands
                    ? createCommands(model, dispatch)
                    : (({} as any) as C);
                return {
                    ...model,
                    ...dispatch,
                    ...commands,
                };
            },
            [model, update, ...memo]
        );

        return vm;
    };
}
