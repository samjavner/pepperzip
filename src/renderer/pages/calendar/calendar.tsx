import * as React from "react";

// VIEW

export const Page: React.FunctionComponent = () => <div>Calendar</div>;
