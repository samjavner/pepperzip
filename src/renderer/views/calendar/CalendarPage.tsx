import * as React from "react";

export const CalendarPage: React.FunctionComponent = () => <div>Calendar</div>;

export default CalendarPage;
