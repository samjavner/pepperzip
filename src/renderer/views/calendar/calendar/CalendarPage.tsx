import * as React from "react";
import { connect } from "react-redux";
import { GlobalState } from "../../../core/model";

export const CalendarPage: React.SFC = props => <div>Calendar</div>;

const mapStateToProps = (state: GlobalState) => ({});

export default connect(mapStateToProps)(CalendarPage);
