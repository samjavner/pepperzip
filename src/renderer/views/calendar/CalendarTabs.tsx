import { Button, ButtonGroup } from "@blueprintjs/core";
import classNames from "classnames";
import * as React from "react";
import { connect } from "react-redux";
import { Dispatch } from "redux";
import * as actions from "../../app/actions";
import { State } from "../../app/state";
import CalendarPage from "./calendar/CalendarPage";

function mapStateToProps(state: State) {
    return {
        activePage: state.activeCalendarPage,
    };
}

function mapDispatchToProps(dispatch: Dispatch<actions.Action>) {
    return {
        onCalendarClick: () =>
            dispatch(actions.setActiveCalendarPage("calendar")),
    };
}

type CalendarTabsProps = ReturnType<typeof mapStateToProps> &
    ReturnType<typeof mapDispatchToProps>;

export const CalendarTabs: React.SFC<CalendarTabsProps> = props => (
    <>
        <div className="pz-navtabs">
            <ButtonGroup minimal={true}>
                <Button
                    className={classNames({
                        "pz-navtabs-button": true,
                        "pz-navtabs-button--active":
                            props.activePage === "calendar",
                    })}
                    icon="calendar"
                    onClick={props.onCalendarClick}
                >
                    Calendar
                </Button>
            </ButtonGroup>
        </div>
        <div style={{ borderTop: "1px solid lightgray" }} />
    </>
);

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(CalendarTabs);
