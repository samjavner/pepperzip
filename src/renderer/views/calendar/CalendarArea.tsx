import * as React from "react";
import { connect } from "react-redux";
import { GlobalState } from "../../core/model";
import * as navigation from "../../core/navigation";
import CalendarPage from "./calendar/CalendarPage";
import CalendarTabs from "./CalendarTabs";

interface CalendarAreaProps {
    activePage: navigation.CalendarPage;
}

export const CalendarArea: React.SFC<CalendarAreaProps> = props => (
    <>
        <CalendarTabs />
        {props.activePage === "calendar" && <CalendarPage />}
    </>
);

const mapStateToProps = (state: GlobalState) => ({
    activePage: navigation.selectors.getActiveCalendarPage(state),
});

export default connect(mapStateToProps)(CalendarArea);
