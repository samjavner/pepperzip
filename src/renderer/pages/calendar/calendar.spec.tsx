import * as React from "react";
import * as renderer from "react-test-renderer";
import * as calendar from "./calendar";

describe("calendar", () => {
    describe("Page", () => {
        it("renders the text 'Calendar'", () => {
            const component = renderer.create(<calendar.Page />);
            const tree = component.toJSON();
            expect(tree).toMatchSnapshot();
        });
    });
});
