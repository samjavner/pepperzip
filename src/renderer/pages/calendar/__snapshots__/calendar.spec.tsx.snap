// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calendar Page renders the text 'Calendar' 1`] = `
<div>
  Calendar
</div>
`;
